src/media/media.ts

```diff
@@ -0,0 +1,820 @@
+    }
+
+    // if a deviceId was passed in, try to use it
+    if (typeof deviceId === 'string') {
+      foundDevice = availableDevices.find((d: MediaDeviceInfo) => d.deviceId === deviceId);
+    }
+
+    // log if we didn't find the requested deviceId
+    if (!foundDevice) {
+      if (typeof deviceId === 'string') {
+        this.sdk.logger.warn(`Unable to find requested ${kind} deviceId`, { deviceId, sessionInfos });
+      }
+
+      // then try to find the sdk default device (if it is not `null`)
+      if (sdkConfigDefault !== null) {
+        foundDevice = availableDevices.find((d: MediaDeviceInfo) => d.deviceId === sdkConfigDefault);
+        // log if we couldn't find the sdk default device
+        if (!foundDevice) {
+          this.sdk.logger.warn(`Unable to find the sdk default ${kind} deviceId`, {
+            deviceId: this.sdk._config.defaults.audioDeviceId,
+            sessionInfos
+          });
+        }
+      }
+    }
+
+    if (!foundDevice) {
+      this.sdk.logger.info(`Using the system default ${kind} device`, { sessionInfos });
+
+      /*
+        SANITY: In chrome an empty string is the system default output device.
+          For mic/camera, we just return `undefined` because gUM will
+          automatically find the default device (the defaults are different
+            between Chrome and FF)
+      */
+      if (kind === 'audiooutput') {
+        foundDevice = { deviceId: '' } as MediaDeviceInfo;
+      }
+    }
+
+    return foundDevice ? foundDevice.deviceId : undefined;
+  }
+
+  /**
+   * Get the current _cached_ media devices
+   */
+  getDevices (): MediaDeviceInfo[] {
+    return this.getState().devices;
+  }
+
+  /**
+   * Get a copy of the current media state
+   */
+  getState (): MediaState {
+    return cloneDeep(this.state);
+  }
+
+  /**
+   * Get the current _cached_ audio devices
+   */
+  getAduioDevices (): MediaDeviceInfo[] {
+    return this.getState().audioDevices;
+  }
+
+  /**
+   * Get the current _cached_ video devices
+   */
+  getVideoDevices (): MediaDeviceInfo[] {
+    return this.getState().videoDevices;
+  }
+
+  /**
+   * Get the current _cached_ output devices
+   */
+  getOutputDevices (): MediaDeviceInfo[] {
+    return this.getState().outputDevices;
+  }
+
+  /**
+   * This will return all active media tracks that 
+   *  were created by the SDK
+   */
+  getAllActiveMediaTracks (): MediaStreamTrack[] {
+    return this.allMediaTracksCreated.filter(track => track.readyState !== 'ended');
+  }
+
+  /**
+   * Look through the cached devices and match based on
+   *  the passed in track's `kind` and `label`. Returns the 
+   *  found device or `undefined` if the device could not be found. 
+   * @param track media stream track with the label to search for
+   */
+  findCachedDeviceByTrackLabel (track?: MediaStreamTrack): MediaDeviceInfo | undefined {
+    if (!track) return;
+    return this.getDevices().find(d => d.label === track.label && d.kind === track.kind);
+  };
```

Comment:
Sorry, I will remove them. I haven't run the linter yet to avoid a massive commit of little changes. I will be sure to lint when I squash the commits.